File: tests/ModelsTests.cs
// =============================================================================
// CyberPatriot Automation Tool - Model Tests
// Author: Maxwell McCormick
// Copyright (c) 2026 Maxwell McCormick. All Rights Reserved.
// =============================================================================

using CyberPatriotAutomation.Core.Models;
using FluentAssertions;
using Xunit;

namespace CyberPatriotAutomation.Tests;

/// <summary>
/// Unit tests for model classes
/// </summary>
public class ModelsTests
{
    #region TaskResult Tests

    [Fact]
    public void TaskResult_ShouldSetExecutedAtOnCreation()
    {
        var result = new TaskResult();

        result.ExecutedAt.Should().BeCloseTo(DateTime.Now, TimeSpan.FromSeconds(5));
    }

    [Fact]
    public void TaskResult_ShouldStoreExecutionResult()
    {
        var result = new TaskResult
        {
            TaskName = "Test Task",
            Success = true,
            Message = "Completed",
            ErrorDetails = null
        };

        result.TaskName.Should().Be("Test Task");
        result.Success.Should().BeTrue();
        result.Message.Should().Be("Completed");
    }

    #endregion

    #region ReadmeData Tests

    [Fact]
    public void ReadmeData_ShouldInitializeWithEmptyCollections()
    {
        var data = new ReadmeData();

        data.Administrators.Should().NotBeNull();
        data.Users.Should().NotBeNull();
        data.ProhibitedSoftware.Should().NotBeNull();
        data.CriticalServices.Should().NotBeNull();
    }

    #endregion

    #region AuthorizedUser Tests

    [Fact]
    public void AuthorizedUser_ShouldStoreUserData()
    {
        var user = new AuthorizedUser
        {
            Username = "testuser",
            IsAdmin = true,
            Password = "SecurePass123!"
        };

        user.Username.Should().Be("testuser");
        user.IsAdmin.Should().BeTrue();
        user.Password.Should().Be("SecurePass123!");
    }

    #endregion

    #region SoftwareRequirement Tests

    [Fact]
    public void SoftwareRequirement_ShouldStoreRequirements()
    {
        var software = new SoftwareRequirement
        {
            Name = "Firefox",
            Version = "latest",
            ShouldBeLatest = true
        };

        software.Name.Should().Be("Firefox");
        software.Version.Should().Be("latest");
        software.ShouldBeLatest.Should().BeTrue();
    }

    #endregion

    #region GroupRequirement Tests

    [Fact]
    public void GroupRequirement_ShouldStoreMembersList()
    {
        var group = new GroupRequirement
        {
            GroupName = "Administrators",
            Members = new List<string> { "admin", "user1" }
        };

        group.GroupName.Should().Be("Administrators");
        group.Members.Should().HaveCount(2);
        group.Members.Should().Contain("admin");
    }

    #endregion

    #region ActionableItem Tests

    [Fact]
    public void ActionableItem_ShouldStoreActionDetails()
    {
        var item = new ActionableItem
        {
            Type = ActionableItemType.CreateGroup,
            Description = "Create backup operators group",
            RawText = "Create a group called BackupOperators"
        };

        item.Type.Should().Be(ActionableItemType.CreateGroup);
        item.Description.Should().Contain("backup");
    }

    [Fact]
    public void ActionableItemType_ShouldHaveAllExpectedTypes()
    {
        var types = Enum.GetValues<ActionableItemType>();

        types.Should().Contain(ActionableItemType.CreateGroup);
        types.Should().Contain(ActionableItemType.CreateUser);
        types.Should().Contain(ActionableItemType.DisableService);
    }

    #endregion
}

File: rust/tests/models_tests.rs
// =============================================================================
// PinnacleCyPat - Model Tests
// =============================================================================

use chrono::Local;
use pinnacle_cypat::models::*;

#[test]
fn task_result_should_set_executed_at_on_creation() {
    let result = TaskResult::default();
    let delta = (Local::now() - result.executed_at).num_seconds().abs();
    assert!(delta < 5);
}

#[test]
fn task_result_should_store_execution_result() {
    let result = TaskResult {
        task_name: "Test Task".to_string(),
        success: true,
        message: "Completed".to_string(),
        error_details: None,
        ..Default::default()
    };
    assert_eq!(result.task_name, "Test Task");
    assert!(result.success);
    assert_eq!(result.message, "Completed");
}

#[test]
fn readme_data_should_initialize_with_empty_collections() {
    let data = ReadmeData::default();
    assert!(data.administrators.is_empty());
    assert!(data.users.is_empty());
    assert!(data.prohibited_software.is_empty());
    assert!(data.critical_services.is_empty());
}

#[test]
fn authorized_user_should_store_user_data() {
    let user = AuthorizedUser {
        username: "testuser".to_string(),
        is_admin: true,
        password: Some("SecurePass123!".to_string()),
        ..Default::default()
    };
    assert_eq!(user.username, "testuser");
    assert!(user.is_admin);
    assert_eq!(user.password.as_deref(), Some("SecurePass123!"));
}

#[test]
fn software_requirement_should_store_requirements() {
    let software = SoftwareRequirement {
        name: "Firefox".to_string(),
        version: Some("latest".to_string()),
        should_be_latest: true,
        ..Default::default()
    };
    assert_eq!(software.name, "Firefox");
    assert_eq!(software.version.as_deref(), Some("latest"));
    assert!(software.should_be_latest);
}

#[test]
fn group_requirement_should_store_members_list() {
    let group = GroupRequirement {
        group_name: "Administrators".to_string(),
        members: vec!["admin".to_string(), "user1".to_string()],
    };
    assert_eq!(group.group_name, "Administrators");
    assert_eq!(group.members.len(), 2);
    assert!(group.members.contains(&"admin".to_string()));
}

#[test]
fn actionable_item_should_store_action_details() {
    let item = ActionableItem {
        item_type: ActionableItemType::CreateGroup,
        description: "Create backup operators group".to_string(),
        raw_text: "Create a group called BackupOperators".to_string(),
        ..Default::default()
    };
    assert_eq!(item.item_type, ActionableItemType::CreateGroup);
    assert!(item.description.contains("backup"));
}

#[test]
fn actionable_item_type_should_have_all_expected_types() {
    // Referencing the variants confirms they exist and are distinct.
    assert_ne!(
        ActionableItemType::CreateGroup,
        ActionableItemType::CreateUser
    );
    assert_ne!(
        ActionableItemType::CreateUser,
        ActionableItemType::DisableService
    );
}
